src/backend/distributed/connection/connection_management.c

```diff
@@ -689,27 +703,40 @@ StartConnectionEstablishment(ConnectionHashKey *key)
 	bool found = false;
 	MultiConnection *connection = NULL;
 	ConnParamsHashEntry *entry = NULL;
+	char **keywords = NULL;
+	char **values = NULL;
+
+	/* if any invalidation happened, reload the configuration */
+	if (!connectionParamHashValid)
+	{
+		DeallocateConnectionParamHash();
+
+		connectionParamHashValid = true;
+	}
+
+	connection = MemoryContextAllocZero(ConnectionContext, sizeof(MultiConnection));
+	connection->internalConnectionContext =
+		AllocSetContextCreate(ConnectionContext, "per connection context",
+							  ALLOCSET_DEFAULT_SIZES);
 
 	/* search our cache for precomputed connection settings */
 	entry = hash_search(ConnParamsHash, key, HASH_ENTER, &found);
-	if (!found || !entry->isValid)
+	if (!found)
 	{
 		/* if they're not found, compute them from GUC, runtime, etc. */
 		GetConnParams(key, &entry->keywords, &entry->values, ConnectionContext);
-
-		entry->isValid = true;
 	}
 
-	connection = MemoryContextAllocZero(ConnectionContext, sizeof(MultiConnection));
-
 	strlcpy(connection->hostname, key->hostname, MAX_NODE_LENGTH);
 	connection->port = key->port;
 	strlcpy(connection->database, key->database, NAMEDATALEN);
 	strlcpy(connection->user, key->user, NAMEDATALEN);
 
+	CopyConnectionParams(connection->internalConnectionContext, &keywords, &values,
+						 entry->keywords, entry->values);
```

Comment:
It seems strange that we're trying to keep this memory for the lifetime of the connection. Do we use keywords/values at all after `PQconnectStartParams`?

Reply:
That's a fair question. Yes, after `PQconnectStartParams ()` we do not need them because they are already [copied](https://github.com/postgres/postgres/blob/554ca6954ecbfc54e7159a609b5ffc7921670185/src/interfaces/libpq/fe-connect.c#L832) to [`PGConn`](https://github.com/postgres/postgres/blob/554ca6954ecbfc54e7159a609b5ffc7921670185/src/interfaces/libpq/fe-connect.c#L720-L727).

I'm undoing that change.